policy: refactor some policy repository functions to act upon a slice of *rule instead of repository itself

ianvernon opened the pull request:
Add a new type to `pkg/policy` called `ruleSlice`, which is wrapper type for `[]*rule`.  This is part of an iterative process to improve policy calculation in Cilium. By making these functions act upon a `ruleSlice`, the functions that apply to the whole policy repository can simply act upon its `rules`; no existing functionality is broken. 

The motivation behind this is so in the future, when policy is computed, we can compute which rules select a given set of labels only once instead of doing it multiple times per policy calculation against that set of labels at each protocol layer (L3, L4, L7, etc). The translation of these rules into the data structures that are utilized within endpoint (`CIDRPolicy`, `L4Policy`) will then be done only against the set of rules that select the endpoint, instead of the against the entire repository. This improves performance because performing the `Match` operation against an endpoint selector is quite costly. 

No functional change is intended by this PR. 

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6265)
<!-- Reviewable:end -->

Commit:
policy: add ruleSlice type

This allows for functions to be written with a slice of *rule as a receiver.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make wildcardL3L4Rules act upon ruleList

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make CanReachIngressRLocked have ruleList receiver

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: change trace function to take a ruleList instead of Repository pointer

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make ResolveL4IngressPolicy have ruleSlice receiver

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make ResolveL4EgressPolicy have ruleSlice receiver

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make ResolveCIDRPolicy have ruleSlice receiver

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: make CanReachEgressRLocked have a ruleSlice receiver

No functional changes intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/policy/repository.go:
```diff
@@ -26,12 +26,16 @@ import (
 
+// ruleSlice is a wrapper around a slice of *rule, which allows for functions
+// to be written with []*rule as a receiver.
+type ruleSlice []*rule
```
Similarly to the `pkg/api/rules.go` create a `pkg/policy/rules.go`

ianvernon replied:
Done.

aanm commented on pkg/policy/repository.go:
```diff
@@ -288,14 +302,10 @@ func (p *Repository) ResolveL4IngressPolicy(ctx *SearchContext) (*L4PolicyMap, e
 		}
 	}
 
-	if result != nil {
```
Maybe this should be kept here as we returned `L4PolicyMap` should contain the revision number when the L4PolicyMap was created, this should not be taken care by the caller of this function.

ianvernon replied:
Done.

aanm commented on pkg/policy/repository.go:
```diff
@@ -349,14 +363,10 @@ func (p *Repository) ResolveL4EgressPolicy(ctx *SearchContext) (*L4PolicyMap, er
-	}
-
-	p.wildcardL3L4Rules(ctx, false, result.Egress)
+	rules.wildcardL3L4Rules(ctx, false, result.Egress)
```
Same here.

ianvernon replied:
Done.

aanm reviewed (dismissed):
I would suggest to move all `ruleSlice` methods to a file called `pkg/policy/rules.go`

Commit:
policy: populate Revision field in resolveL4{Egress,Ingress}policy

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: move ruleSlice functions to new rules.go

No functional change is intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

nebril reviewed (approved):
LGTM
